File: NET/JNetProto.Sharp/DataTypes.cs
using System;
using System.Collections;

namespace JNetProto.Sharp
{
    public static class DataTypes
    {
        public static Type GetClass(DataType type)
        {
            switch (type)
            {
                case DataType.Bool: return typeof(bool);
                case DataType.I8: return typeof(byte);
                case DataType.I16: return typeof(short);
                case DataType.I32: return typeof(int);
                case DataType.I64: return typeof(long);
                case DataType.F32: return typeof(float);
                case DataType.F64: return typeof(double);
                case DataType.F128: return typeof(decimal);
                case DataType.Char: return typeof(char);
                case DataType.UTF8: return typeof(string);
                case DataType.Duration: return typeof(TimeSpan);
                case DataType.Timestamp: return typeof(DateTime);
                case DataType.Guid: return typeof(Guid);
                default: throw new ArgumentException(type.ToString());
            }
        }

        public interface IDataType { DataType Kind { get; } }
        public record SingleDt(DataType Kind) : IDataType;
        public record ArrayDt(DataType Kind, int Rank, IDataType Item) : IDataType;
        public record MapDt(DataType Kind, IDataType Key, IDataType Val) : IDataType;

        public static IDataType GetKind(object instance)
        {
            var type = instance as Type ?? instance.GetType();
            if (type.IsArray)
            {
                var rank = type.GetArrayRank();
                var item = type.GetElementType();
                return new ArrayDt(DataType.Array, rank, GetKind(item));
            }
            if (type.IsAssignableTo(typeof(IDictionary)))
            {
                var dict = (IDictionary)instance;
                var f = default(DictionaryEntry);
                foreach (DictionaryEntry entry in dict)
                {
                    f = entry;
                    break;
                }
                return new MapDt(DataType.Map, GetKind(f.Key), GetKind(f.Value));
            }
            return new SingleDt(GetSingleKind(type));
        }

        private static DataType GetSingleKind(Type type)
        {
            switch (type.FullName)
            {
                case "System.Boolean": return DataType.Bool;
                case "System.Byte": return DataType.I8;
                case "System.Int16": return DataType.I16;
                case "System.Int32": return DataType.I32;
                case "System.Int64": return DataType.I64;
                case "System.Single": return DataType.F32;
                case "System.Double": return DataType.F64;
                case "System.Decimal": return DataType.F128;
                case "System.Char": return DataType.Char;
                case "System.String": return DataType.UTF8;
                case "System.TimeSpan": return DataType.Duration;
                case "System.DateTime": return DataType.Timestamp;
                case "System.Guid": return DataType.Guid;
                default: throw new ArgumentException(type.ToString());
            }
        }
    }
}

File: JVM/jnetproto-java/src/main/java/jnetproto/java/DataTypes.java
package jnetproto.java;

import java.math.BigDecimal;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

import org.javatuples.Tuple;

import jnetproto.java.compat.Reflect;

public final class DataTypes {
    public static Class<?> getClass(DataType type) {
        switch (type) {
            case Bool: return boolean.class;
            case I8: return byte.class;
            case I16: return short.class;
            case I32: return int.class;
            case I64: return long.class;
            case F32: return float.class;
            case F64: return double.class;
            case F128: return BigDecimal.class;
            case Char: return char.class;
            case UTF8: return String.class;
            case Duration: return Duration.class;
            case Timestamp: return LocalDateTime.class;
            case Guid: return UUID.class;
            default: throw new IllegalArgumentException(type.toString());
        }
    }

    public interface IDataType { DataType Kind(); }
    private record SingleDt(DataType Kind) implements IDataType { };
    public record ArrayDt(DataType Kind, int Rank, IDataType Item) implements IDataType {};
    public record MapDt(DataType Kind, IDataType Key, IDataType Val) implements IDataType {};
    public record ListDt(DataType Kind, IDataType Item) implements IDataType {};

    public static IDataType getKind(Object instance)
    {
        var type = instance instanceof Class<?> cl ? cl : instance.getClass();
        if (type.isArray())
        {
            var item = type.getComponentType();
            var rank = Reflect.getRank(type);
            if (rank == 1)
            {
                if (item == Object.class)
                    return new SingleDt(DataType.Bag);
                if (item == byte.class)
                    return new SingleDt(DataType.Binary);
            }
            return new ArrayDt(DataType.Array, rank, getKind(item));
        }
        if (Tuple.class.isAssignableFrom(type))
        {
            return new SingleDt(DataType.Tuple);
        }
        if (Map.class.isAssignableFrom(type))
        {
            var dict = (Map<?,?>)instance;
            Map.Entry<?,?> f = null;
            for (var entry : dict.entrySet())
            {
                f = (Map.Entry<?,?>)entry;
                break;
            }
            return new MapDt(DataType.Map, getKind(f.getKey()), getKind(f.getValue()));
        }
        if (Set.class.isAssignableFrom(type))
        {
            var item = ((Set<?>)instance).iterator().next();
            return new ListDt(DataType.Set, getKind(item));
        }
        if (List.class.isAssignableFrom(type))
        {
            var item = ((List<?>)instance).get(0);
            return new ListDt(DataType.List, getKind(item));
        }
        return new SingleDt(getSingleKind(type));
    }

    private static DataType getSingleKind(Class<?> type) {
        switch (type.getName()) {
            case "boolean": case "java.lang.Boolean": return DataType.Bool;
            case "byte": case "java.lang.Byte": return DataType.I8;
            case "short": case "java.lang.Short": return DataType.I16;
            case "int": case "java.lang.Integer": return DataType.I32;
            case "long": case "java.lang.Long": return DataType.I64;
            case "float": case "java.lang.Float": return DataType.F32;
            case "double": case "java.lang.Double": return DataType.F64;
            case "java.math.BigDecimal": return DataType.F128;
            case "char": case "java.lang.Character": return DataType.Char;
            case "java.lang.String": return DataType.UTF8;
            case "java.time.Duration": return DataType.Duration;
            case "java.time.LocalDateTime": return DataType.Timestamp;
            case "java.util.UUID": return DataType.Guid;
            default: throw new IllegalArgumentException(type.toString());
        }
    }
}